### ISSUE TITLE
Karma firefox test failing with `Cannot find module 'Firefox.app/Contents/MacOS/firefox'`

### ISSUE THREAD

**Issue by gregmagolan** [bazelbuild/rules_typescript#296]
After the last update of rules_webtesting in rules_typescript, `ts_web_test_suite` targets that test against the `@io_bazel_rules_webtesting//browsers:firefox-local` browser fail with `Cannot find module 'Firefox.app/Contents/MacOS/firefox'`.

This is true for OSX at least. I haven't checked if it fails on Linux or Windows.

//cc @filipesilva @alexeagle

**Comment by alexeagle** [comment]
Fails on Linux too.

**Comment by filipesilva** [comment]
Does this happen only on local machines? CI seems to be green.

**Comment by gregmagolan** [comment]
> Does this happen only on local machines? CI seems to be green.

The firefox tests have been commented out on CI

**Comment by filipesilva** [comment]
Oh that's right. On the Windows CI they seem to be running though: https://github.com/bazelbuild/rules_typescript/blob/master/.bazelci/presubmit.yml

If both firefox and chrome tests are commented out for linux/osx, does that mean that `ts_web_test` is not being tested there at all? I think those are the only browsers that are available.

**Comment by gregmagolan** [comment]
Chrome is still tested in CircleCI linux at least. But yes, I think we need to get more testing up for these.

**Comment by gregmagolan** [comment]
A little more information on this. rules_webtesting provides a path to the chromium executable in the test environment:

```
{
  "capabilities": {
    "browserName": "chromium",
    "goog:chromeOptions": {
      "args": [
        "--headless",
        "--use-gl=swiftshader-webgl"
      ],
      "binary": "%FILE:CHROMIUM%"
    },
    "google:wslConfig": {
      "args": [
        "--port=%WSLPORT:WSL%"
      ],
      "binary": "%FILE:CHROMEDRIVER%",
      "port": "%WSLPORT:WSL%",
      "shutdown": true,
      "status": true
    }
  },
  "environment": "local",
  "browserLabel": "@io_bazel_rules_webtesting//browsers:chromium-local",
  "webTestFiles": [
    {
      "namedFiles": {
        "CHROMEDRIVER": "io_bazel_rules_webtesting/third_party/chromedriver/chromedriver.out/chromedriver",
        "CHROMIUM": "io_bazel_rules_webtesting/third_party/chromium/chromium.out/chrome-mac/Chromium.app/Contents/MacOS/chromium",
        "WEBDRIVER_SERVER_LIGHT": "io_bazel_rules_webtesting/go/wsl/main/darwin_amd64_stripped/main"
      }
    }
  ],
  "extension": {}
}
```

but for Firefox it now provides an `archiveFile` and the path for firefox that is provided is for the archive contents. This change is what broke the firefox tests.

There is a `io_bazel_rules_webtesting/web/internal/extract.sh` file in the runfiles for the firefox test which isn't in the chromium runfiles. Is the test runner meant to run extract.sh on the firefox archive now?

**Comment by gregmagolan** [comment]
@alexeagle @filipesilva Thoughts?

**Comment by filipesilva** [comment]
Extracting the firefox archive and providing a final path for the executable sounds like that kind of thing that should be happening inside `rules_webtesting`. I don't think we should be trying to extract internal `rules_webtesting` downloaded archives and processing their files.

**Comment by gregmagolan** [comment]
So it looks like this rules_webtesting added a feature where the archive can be extracted at either build time or runtime: https://github.com/bazelbuild/rules_webtesting/pull/222.

And firefox was change to be extracted at `run` here: https://github.com/bazelbuild/rules_webtesting/pull/293

Later a comment got added that Firefox needs to be in a writeable directory (I assume that means that it must be extracted in the runfiles directory when the test runs):

```
web_test_archive(
    name = "firefox",
    archive = "@org_mozilla_firefox//file",
    # Firefox needs to be in a writable directory.
    extract = "run",
```

**Comment by gregmagolan** [comment]
Some more context:

```
    doc = """Specifies an archive file with named files in it.

        If extract=="run", then the archive will only be extracted if WTL wants one
        of the named files in it.""",
```

The extract is handled automatically by the .go code but in our case we may have to handle it ourselves.

**Comment by gregmagolan** [comment]
Fixed with #299
